### PR TITLE
OpenRV integration

### PR DESCRIPTION
**Pull request by ddesmond**
## Changelog Description
Draft OpenRV integration including Ftrack support
Features:
- View FTrack action, Review Ftrack action
- OpenPype Menus package
- OpenPype Comments package
- Loading and Publising annotations 

## Additional info

You need to compile OpenRV first, head to OpenRV github and do it!

### TODO

> - to have a review session you need a task in ftrack where RV can hook to  -> in my case ive setup/hardcoded 
> ```
> task_name = "prepDaily"
> asset_name = "DaliesPrep"
> ```
> so RV can have its session files stored with a proper task. this should be somehow parametrized ?

Roy: Untested - still todo.

**TODO**

- [ ] it needs a proper rv home env var setup otherwise it wont work. _Roy: Bit unclear where this needs to be used and why it's required, but something to pay attention to._
- [ ] Have OP menus working stabily by serving a customSUPPORT path and config path or to install the packages automatically. _Roy: This is currently hardcoded into the OP codebase (in launch prehook) and should be resolved. Relevant todo is in the code. (there should be better ways to automate this)_
- [x] OpenRV needs a clean pythonpath and we have to rebuild everything afterwards. _Roy: This has currently been worked around with a hack using `sys.path` in OpenPype menus package._
- [x] Fix remainder of Hound comments
- [x] Test whether everything works with OpenPype on PYTHONPATH - according to @ddesmond it breaks with OP on path but I didn't have that issue. He mentioned it's something with OTIO / OCIO or alike, so I'll have to investigate those features.
- [x] Fix loading frame ranges for sequences
- [x] Fix removing loaded containers
- [x] Add support for applying correct OCIO color space per loaded footage using the representation's colorspace.
- [x] Fix schema refactor for imageio settings
   - [ ] Update schema when https://github.com/ynput/OpenPype/pull/4700 is merged.
- [ ] Test the annotations/commenting workflow
- [ ] Styling of the OpenPype Qt widgets is slighty off (non-critical), see image below
    - [ ] text in header is small 
    - [ ] spacing in column/row highlighting
    - [ ] scrollbars look different?

afbeelding

### Testing notes

1. Run OpenRV through OpenPype and go crazy
2. Test the ftrack actions to see whether they do what they should

### CODE REVIEW

File: openpype/hosts/openrv/plugins/load/load_frames.py [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+        # Command fails on unicode so we must force it to be strings
+        filepath = str(filepath)
+
+        node_name = "{}_{}".format(namespace, name) if namespace else name
+        namespace = namespace if namespace else context["asset"]["name"]
+
+        set_session_fps()
+        reset_frame_range()
+        loaded_node = rv.commands.addSourceVerbose([filepath])
+        print("loaded_node", loaded_node)
```
Comment (BigRoy) [inline openpype/hosts/openrv/plugins/load/load_frames.py:34]:
```suggestion
```

OR

```suggestion
        self.log.debug("Loaded node: {}".format(loaded_node))
```

File: openpype/hosts/openrv/plugins/load/load_frames.py [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+        rv.commands.setSourceMedia(node, [filepath])
+        rv.commands.setStringProperty(node + ".media.name", ["newname"], True)
+        rv.commands.setStringProperty(node + ".media.repName", ["repname"], True)
+        rv.commands.setStringProperty(node + ".openpype.representation", [str(representation["_id"])], True)
+
+
+
+    def remove(self, container):
+        node = container["node"]
+        # not implemented
```
Comment (BigRoy) [inline openpype/hosts/openrv/plugins/load/load_frames.py:61]:
```suggestion
        # todo: not implemented
```

File: openpype/hosts/openrv/startup/pkgs_source/comments/PACKAGE [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+requires: ''
+optional: true
+
+modes: 
+  - file: comments
+    load: immediate
+
+description: |
+
+  <p> Adds COmments to OpenRV </p>
```
Comment (BigRoy) [inline openpype/hosts/openrv/startup/pkgs_source/comments/PACKAGE:16]:
```suggestion
  <p> Adds Comments to OpenRV </p>
```

Should this and maybe the actual folder mention actually mention **OpenPype** somewhere?

File: openpype/settings/entities/schemas/projects_schema/schemas/schema_openrv_publish.json [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+    {
+      "type": "schema_template",
+      "name": "template_publish_plugin",
+      "template_data": [
+          {
+              "key": "ValidateContainers",
+              "label": "ValidateContainers"
+          }
+      ]
+    }
```
Comment (BigRoy) [inline openpype/settings/entities/schemas/projects_schema/schemas/schema_openrv_publish.json:48]: Is this used for anything? Since the ValidateContainers plug-in doesn't appear in this PR?

Same goes for the `ValidateWorkfilePaths` above. Seems like a straight up copy from Houdini maybe and shouldn't actually be in here?

File: openpype/hosts/openrv/addon.py [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+class OpenRVAddon(OpenPypeModule, IHostAddon):
+    name = "openrv"
+    host_name = "openrv"
+
+    def initialize(self, module_settings):
+        self.enabled = True
+
+    def add_implementation_envs(self, env, app):
+        """Modify environments to contain all required for implementation."""
+        sys.path.insert(0, os.path.join(os.getenv("OPENPYPE_ROOT"), "vendor", "python"))
```
Comment (BigRoy) [inline openpype/hosts/openrv/addon.py:18]: Why is this needed?

File: openpype/hosts/openrv/api/commands.py [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+    asset_doc = get_current_project_asset()
+    asset_data = asset_doc["data"]
+    fps = float(asset_data.get("fps", 25))
+    rv.commands.setFPS(fps)
+
+
+
+def create_support_ticket():
+    import webbrowser
+    webbrowser.open("http://localhost:5400/tickets/create_tickets")
```
Comment (BigRoy) [inline openpype/hosts/openrv/api/commands.py:43]: This seems very specific to something custom to you? Or is this a default port openRV starts through which you can create tickets?

**Review by BigRoy — COMMENTED** [review]
Great stuff. There are quite a few cosmetic things the hound trips over. Would you like me to clean that up for you? I can push directly into your branch (if you'd like that) or create a PR to your repo if you prefer that?

**Comment by ddesmond** [comment]
> Great stuff. There are quite a few cosmetic things the hound trips over. Would you like me to clean that up for you? I can push directly into your branch (if you'd like that) or create a PR to your repo if you prefer that?

push directly to my branch

File: openpype/hosts/openrv/hooks/pre_debughooks.py
Comment (BigRoy) [inline openpype/hosts/openrv/hooks/pre_debughooks.py:1]: This **debug file** can be completely removed, correct?

File: openpype/hosts/openrv/plugins/create/create_debug.py
Comment (BigRoy) [inline openpype/hosts/openrv/plugins/create/create_debug.py:1]: This **debug file** can be completely removed, correct?

File: openpype/settings/entities/schemas/projects_schema/schemas/schema_openrv_publish.json (hunk 4 above)
Comment (ddesmond) [reply openpype/settings/entities/schemas/projects_schema/schemas/schema_openrv_publish.json:48]: yes, can be removed. there will be more leftovers needed for cleanup for sure...

File: openpype/hosts/openrv/api/commands.py (hunk 6 above)
Comment (ddesmond) [reply openpype/hosts/openrv/api/commands.py:41]: ah, i thought i removed that one, you can remove it, its for opening tickets in the future for the addon, like githuub or so

File: openpype/hosts/openrv/hooks/pre_debughooks.py
Comment (ddesmond) [reply openpype/hosts/openrv/hooks/pre_debughooks.py:1]: sure!

File: openpype/hosts/openrv/plugins/create/create_debug.py
Comment (ddesmond) [reply openpype/hosts/openrv/plugins/create/create_debug.py:1]: sure

File: openpype/hosts/openrv/hooks/pre_config_setup.py
Comment (BigRoy) [inline openpype/hosts/openrv/hooks/pre_config_setup.py:1]: It looks like this launch hook only tries to set the OCIO config related settings. Can this basically be replaced by what [Fusion OCIO prelaunch hook](https://github.com/ynput/OpenPype/blob/develop/openpype/hosts/fusion/hooks/pre_fusion_ocio_hook.py) does? 

That would require fixing up how the OCIO color settings are structured. They currently differ in structure from other hosts and might be best to align with this PR: https://github.com/ynput/OpenPype/pull/4700 - what do you think @ddesmond @jakubjezek001 ?

Comment (ddesmond) [reply openpype/hosts/openrv/hooks/pre_config_setup.py:1]: whatever works best in grand scheme of things.

File: openpype/hosts/openrv/hooks/pre_ftrackdata.py
Comment (BigRoy) [inline openpype/hosts/openrv/hooks/pre_ftrackdata.py:1]: @ddesmond I can't find any logic which uses this - did you forget to commit certain things that use this? I can't find anything using the `OPENPYPE_LOADER_REPRESENTATIONS` variable.

Any ideas?

Comment (ddesmond) [reply openpype/hosts/openrv/hooks/pre_ftrackdata.py:1]: it was in the class PreFtrackData(PreLaunchHook): you removed maybe

Comment (BigRoy) [reply openpype/hosts/openrv/hooks/pre_ftrackdata.py:1]: It's still there, but it's only setting the value. I don't see where that env variable is being picked up to do anything useful. :)

Comment (ddesmond) [reply openpype/hosts/openrv/hooks/pre_ftrackdata.py:1]: sec, ill push an update

Comment (ddesmond) [reply openpype/hosts/openrv/hooks/pre_ftrackdata.py:1]: pushed, check the menu py file from the rv startup folder

Comment (BigRoy) [reply openpype/hosts/openrv/hooks/pre_ftrackdata.py:1]: Thanks! looks good.

**Comment by BigRoy** [comment]
I've spent today testing the integration but mostly cleaning up the PR code. I had some issues getting the integration to work and setup but I believe with the commits I added now you should be able to just point to the right `openrv` executable and get it working right of the bat.

The workfiles tool works, I was able to load frames with the loader (however it'd only ever load 1 frame of an exr sequence instead of the whole sequence - I will need to investigate more).
I haven't looked into any of the "annotation" features since it seemed heavily coupled to Ftrack which I haven't tested today.

File: openpype/hosts/openrv/startup/pkgs_source/comments/comments.py [hunk 7]
```diff
@@ -0,0 +1,313 @@
...
+                else:
+                    for y,num in enumerate(frames_in, start=set_start):
+                        return num
+            else:
+                if do == "next":
+                    next_frame = min([i for i in frames_in if frame < i])
+                    return next_frame
+                else:
+                    prev_frame = max([i for i in frames_in if frame > i])
+                    return prev_frame
```
Comment (BigRoy) [inline openpype/hosts/openrv/startup/pkgs_source/comments/comments.py]: @ddesmond Can you explain what this code is trying to do?

**Comment by BigRoy** [comment]
I was able to reproduce some of the `PYTHONPATH` conflict issues @ddesmond described with OpenPype and running OpenRV without clearing (or at least removing a lot from `PYTHONPATH`).

Here's an error with OCIO when trying to activate a colorspace:
```
ERROR: Problem occurred while loading OCIO settings for OCIOFile: module 'PyOpenColorIO' has no attribute 'Constants'
```

@iLLiCiTiT how have we been avoiding similar conflicts in other packages? Did we just get lucky with having somewhat compatible python packages?

**Comment by BigRoy** [comment]
Here's a screen capture of the current state of this PR with some notes as a ramble on about it and misclick half of the UI buttons the whole time.

https://user-images.githubusercontent.com/2439881/232827997-9cdef99a-c5e6-4635-8353-d7f3b147ae18.mp4

File: openpype/hosts/openrv/addon.py [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+
+    def initialize(self, module_settings):
+        self.enabled = True
+
+    def add_implementation_envs(self, env, app):
+        """Modify environments to contain all required for implementation."""
+        vendor_python = os.path.join(os.getenv("OPENPYPE_ROOT"),
+                                     "vendor",
+                                     "python")
+        sys.path.insert(0, vendor_python)
```
Comment (BigRoy) [inline openpype/hosts/openrv/addon.py:21]: I'm not entirely sure this is needed actually.

File: openpype/hosts/openrv/api/commands.py [hunk 9]
```diff
@@ -0,0 +1,36 @@
...
+    asset_doc = get_current_project_asset()
+    asset_data = asset_doc["data"]
+
+    frame_start = int(asset_data.get(
+        "frameStart",
+        asset_data.get("edit_in")))
+
+    frame_end = int(asset_data.get(
+        "frameEnd",
+        asset_data.get("edit_out")))
```
Comment (BigRoy) [inline openpype/hosts/openrv/api/commands.py:21]: This should go without the edit_in and edit_out fallbacks

File: openpype/hosts/openrv/api/ocio.py [hunk 10]
```diff
@@ -0,0 +1,115 @@
+"""Helper functions to apply OCIO colorspace settings on groups.
+
+This tries to set the relevant OCIO settings on the group's look and render
+pipeline similar to what the OpenColorIO Basic Color Management package does in
+OpenRV through its `ocio_source_setup` python file.
```
Comment (BigRoy) [inline openpype/hosts/openrv/api/ocio.py:5]: It would be great if we could also autoload this package - but I know too little of RV to say what the best approach is to do so.

File: openpype/hosts/openrv/api/pipeline.py [hunk 11]
```diff
@@ -0,0 +1,245 @@
...
+
+    def get_containers(self):
+        for container in get_containers():
+            yield container
+
+    def update_context_data(self, data, changes):
+        imprint("root", data, prefix=OPENPYPE_ATTR_PREFIX)
+
+    def get_context_data(self):
+        return read("root", prefix=OPENPYPE_ATTR_PREFIX)
```
Comment (BigRoy) [inline openpype/hosts/openrv/api/pipeline.py:81]: I'm now storing global data on root since that node name sounded sensible, but I'm not sure if that's what its intent should be. Maybe someone more knowledgeable of (open)RV can comment on whether that is good practice.

File: openpype/hosts/openrv/api/review.py
Comment (BigRoy) [inline openpype/hosts/openrv/api/review.py:1]: These functions are basically untested at this stage as far as I know.

File: openpype/hosts/openrv/hooks/pre_setup_openrv.py [hunk 12]
```diff
@@ -0,0 +1,55 @@
...
+                        shutil.copy(filepath, startup_python)
+
+        # TODO: Make sure we don't override a full studios RV_SUPPORT_PATH
+        print("Setting RV_SUPPORT_PATH", startup)
+        self.launch_context.env["RV_SUPPORT_PATH"] = str(startup)
+
+        # TODO: OpenRV does write files into RV_SUPPORT_PATH during runtime
+        #   so we should actually not deploy that as a path inside OP deploy
+        #   to avoid openpype checksum validation issues, etc. but for now
+        #   it helps running from code to run the integration for development
```
Comment (BigRoy) [inline openpype/hosts/openrv/hooks/pre_setup_openrv.py:55]: I'd say these todos are critical to resolve before merging.

File: openpype/hosts/openrv/plugins/publish/extract_annotations.py
Comment (BigRoy) [inline openpype/hosts/openrv/plugins/publish/extract_annotations.py:1]: Pretty much haven't worked on this extractor at all - it's clearly WIP still.

**Review by BigRoy — COMMENTED** [review]
Some comments to highlight todos.

File: openpype/hosts/openrv/hooks/pre_setup_openrv.py (hunk 12 above)
Comment (BigRoy) [reply openpype/hosts/openrv/hooks/pre_setup_openrv.py]: This has now been somewhat resolved by installing to a temporary directory instead

File: openpype/hosts/openrv/api/ocio.py (hunk 10 above)
Comment (BigRoy) [reply openpype/hosts/openrv/api/ocio.py:5]: I've asked more information about this on the Shotgrid Software community about [Autoload OpenColorIO Basic Color Management RV package](https://community.shotgridsoftware.com/t/autoload-opencolorio-basic-color-management-rv-package/17239/1).

File: openpype/hosts/openrv/hooks/pre_setup_openrv.py (hunk 12 above)
Comment (BigRoy) [reply openpype/hosts/openrv/hooks/pre_setup_openrv.py]: I've asked for other recommended solutions on the Shotgrid Software community's OpenRV subforum here in [Launching RV with plugins from network path and force loaded](https://community.shotgridsoftware.com/t/launching-rv-with-plugins-from-network-path-and-force-loaded/17236).

**Comment by rontown** [comment]
Interesting to see you successfully compiled a version of OpenRV. 
Ive been trying on and off to compile it on windows 10 and rocky 9.2 for months to no avail. Even more exciting to see the developing integration with openpype/ayon!

**Comment by antirotor** [comment]
> Interesting to see you successfully compiled a version of OpenRV.
> Ive been trying on and off to compile it on windows 10 and rocky 9.2 for months to no avail. Even more exciting to see the developing integration with openpype/ayon!

Building on Windows does work, just follow closely this guide - https://github.com/AcademySoftwareFoundation/OpenRV/discussions/91

File: openpype/hosts/openrv/hooks/pre_setup_openrv.py [hunk 13]
```diff
@@ -0,0 +1,64 @@
...
+    """Pre-hook for openrv"""
+    app_groups = ["openrv"]
+
+    def execute(self):
+
+        executable = self.application.find_executable()
+        if not executable:
+            self.log.error("Unable to find executable for RV.")
+            return
+
```
Comment (iLLiCiTiT) [inline openpype/hosts/openrv/hooks/pre_setup_openrv.py:21]: It would not get to prelaunch hook if executable would not be found.
```suggestion
```

Comment (BigRoy) [reply openpype/hosts/openrv/hooks/pre_setup_openrv.py:21]: The executable path is still needed for the logic below - so I guess it's just the `if` statement that you want removed, not the `find_executable?` Or is there a better way in the prelaunch hook to get the executable path?

Comment (BigRoy) [reply openpype/hosts/openrv/hooks/pre_setup_openrv.py:21]: For what it's worth and for your information:

This PR was originally opened by @ddesmond

I contributed into the PR at the time because I could because I had write permissions to the repo - I currently don't anymore so I am personally unable to contribute to this PR.

I believe @ddesmond is not working on this feature currently meaning that without other contributions from someone with write access this will end up staying as a stale PR.

Comment (ddesmond) [reply openpype/hosts/openrv/hooks/pre_setup_openrv.py:21]: If you need permissions i can give some to you @BigRoy

Comment (ddesmond) [reply openpype/hosts/openrv/hooks/pre_setup_openrv.py:21]: there is also https://github.com/ddesmond/OpenPype/pull/8
You have been added as a colab to my repo @BigRoy , let me know guys if you need more help

**Comment by kalisp** [comment]
Closing this one in favor of https://github.com/ynput/OpenPype/pull/6185 - which cleans up a bit this one, translates it into Ayon addon.

Any comments welcomed in that PR or in Issues in https://github.com/ynput/ayon-openrv
